In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import requests
import os
import json
import tweepy
%matplotlib inline
%config InlineBackend.figure_format = 'retina'

In [2]:
# download image prediction tsv file programatically
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
response = requests.get(url)
response

<Response [200]>

In [3]:
# save image prediction file in project folder
with open('image-predictions.tsv', mode='wb') as file:
    file.write(response.content)

Assign Image Predictions file to a Dataframe and have a quick look if it worked:

In [4]:
df1 = pd.read_csv('image-predictions.tsv', sep='\s+')
df1.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

In [5]:
df1.tweet_id.count()

2075

Again, import the twitter archive file and quickly check the output:

In [6]:
df2 = pd.read_csv('twitter-archive-enhanced.csv')
df2.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

In [7]:
df2.tweet_id.count()

2356

In [37]:
# authorize tweepy

consumer_key = 'hidden'
consumer_secret = 'hidden'
access_token = 'hidden'
access_secret = 'hidden'

auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

api = tweepy.API(auth, wait_on_rate_limit=True, wait_on_rate_limit_notify=True)

In [9]:
test_tweet = api.get_status(df2.tweet_id[1000], tweet_mode='extended')

In [10]:
test_tweet._json

{'created_at': 'Wed Jun 29 01:23:16 +0000 2016',
 'id': 747963614829678593,
 'id_str': '747963614829678593',
 'full_text': 'PUPPER NOOOOO BEHIND YOUUU 10/10 pls keep this pupper in your thoughts https://t.co/ZPfeRtOX0Q',
 'truncated': False,
 'display_text_range': [0, 70],
 'entities': {'hashtags': [],
  'symbols': [],
  'user_mentions': [],
  'urls': [],
  'media': [{'id': 747963600220917761,
    'id_str': '747963600220917761',
    'indices': [71, 94],
    'media_url': 'http://pbs.twimg.com/media/CmFM7ngXEAEitfh.jpg',
    'media_url_https': 'https://pbs.twimg.com/media/CmFM7ngXEAEitfh.jpg',
    'url': 'https://t.co/ZPfeRtOX0Q',
    'display_url': 'pic.twitter.com/ZPfeRtOX0Q',
    'expanded_url': 'https://twitter.com/dog_rates/status/747963614829678593/photo/1',
    'type': 'photo',
    'sizes': {'thumb': {'w': 150, 'h': 150, 'resize': 'crop'},
     'medium': {'w': 937, 'h': 632, 'resize': 'fit'},
     'small': {'w': 680, 'h': 459, 'resize': 'fit'},
     'large': {'w': 937, 'h': 632, '

In [11]:
data = {}
data['tweet'] = []
tweet_count = 1
tweet_errors = {}

for tweet_id in df2.tweet_id.values:
    try: 
        tweet = api.get_status(tweet_id, tweet_mode='extended') 
        print('Tweet Count: {}'.format(tweet_count))
        print('Tweet ID: {}'.format(tweet_id))
        
        # get info for retweet count, favorite count and followers count
        json_tweet = tweet._json
        retweet_count = json_tweet['retweet_count']
        favorite_count = json_tweet['favorite_count']
        followers_count = json_tweet['user']['followers_count']
        
        # append the infos to the data dict.
        data['tweet'].append({'tweet_id': tweet_id,
                              'retweet_count': retweet_count, 
                              'favorite_count': favorite_count, 
                              'followers_count': followers_count
                              })
        
        tweet_count += 1
         
        # catch errors in tweet errors dict.
    except Exception as e:
            print('An Error occured on ID: ' + str(tweet_id) + 'Error Message: ' + str(e))
            tweet_errors[str(tweet_id)] = json_tweet

Tweet Count: 1
Tweet ID: 892420643555336193
Tweet Count: 2
Tweet ID: 892177421306343426
Tweet Count: 3
Tweet ID: 891815181378084864
Tweet Count: 4
Tweet ID: 891689557279858688
Tweet Count: 5
Tweet ID: 891327558926688256
Tweet Count: 6
Tweet ID: 891087950875897856
Tweet Count: 7
Tweet ID: 890971913173991426
Tweet Count: 8
Tweet ID: 890729181411237888
Tweet Count: 9
Tweet ID: 890609185150312448
Tweet Count: 10
Tweet ID: 890240255349198849
Tweet Count: 11
Tweet ID: 890006608113172480
Tweet Count: 12
Tweet ID: 889880896479866881
Tweet Count: 13
Tweet ID: 889665388333682689
Tweet Count: 14
Tweet ID: 889638837579907072
Tweet Count: 15
Tweet ID: 889531135344209921
Tweet Count: 16
Tweet ID: 889278841981685760
Tweet Count: 17
Tweet ID: 888917238123831296
Tweet Count: 18
Tweet ID: 888804989199671297
Tweet Count: 19
Tweet ID: 888554962724278272
An Error occured on ID: 888202515573088257Error Message: [{'code': 144, 'message': 'No status found with that ID.'}]
Tweet Count: 20
Tweet ID: 88807843445

Tweet Count: 167
Tweet ID: 858843525470990336
Tweet Count: 168
Tweet ID: 858471635011153920
Tweet Count: 169
Tweet ID: 858107933456039936
Tweet Count: 170
Tweet ID: 857989990357356544
Tweet Count: 171
Tweet ID: 857746408056729600
Tweet Count: 172
Tweet ID: 857393404942143489
Tweet Count: 173
Tweet ID: 857263160327368704
Tweet Count: 174
Tweet ID: 857214891891077121
Tweet Count: 175
Tweet ID: 857062103051644929
Tweet Count: 176
Tweet ID: 857029823797047296
Tweet Count: 177
Tweet ID: 856602993587888130
Tweet Count: 178
Tweet ID: 856543823941562368
Tweet Count: 179
Tweet ID: 856526610513747968
Tweet Count: 180
Tweet ID: 856330835276025856
Tweet Count: 181
Tweet ID: 856288084350160898
Tweet Count: 182
Tweet ID: 856282028240666624
Tweet Count: 183
Tweet ID: 855862651834028034
Tweet Count: 184
Tweet ID: 855860136149123072
Tweet Count: 185
Tweet ID: 855857698524602368
Tweet Count: 186
Tweet ID: 855851453814013952
Tweet Count: 187
Tweet ID: 855818117272018944
Tweet Count: 188
Tweet ID: 8554594

Tweet Count: 338
Tweet ID: 831926988323639298
Tweet Count: 339
Tweet ID: 831911600680497154
Tweet Count: 340
Tweet ID: 831670449226514432
Tweet Count: 341
Tweet ID: 831650051525054464
Tweet Count: 342
Tweet ID: 831552930092285952
Tweet Count: 343
Tweet ID: 831322785565769729
Tweet Count: 344
Tweet ID: 831315979191906304
Tweet Count: 345
Tweet ID: 831309418084069378
Tweet Count: 346
Tweet ID: 831262627380748289
Tweet Count: 347
Tweet ID: 830956169170665475
Tweet Count: 348
Tweet ID: 830583320585068544
Tweet Count: 349
Tweet ID: 830173239259324417
Tweet Count: 350
Tweet ID: 830097400375152640
Tweet Count: 351
Tweet ID: 829878982036299777
Tweet Count: 352
Tweet ID: 829861396166877184
Tweet Count: 353
Tweet ID: 829501995190984704
Tweet Count: 354
Tweet ID: 829449946868879360
Tweet Count: 355
Tweet ID: 829374341691346946
Tweet Count: 356
Tweet ID: 829141528400556032
Tweet Count: 357
Tweet ID: 829011960981237760
Tweet Count: 358
Tweet ID: 828801551087042563
Tweet Count: 359
Tweet ID: 8287703

Tweet Count: 512
Tweet ID: 809808892968534016
Tweet Count: 513
Tweet ID: 809448704142938112
Tweet Count: 514
Tweet ID: 809220051211603969
Tweet Count: 515
Tweet ID: 809084759137812480
Tweet Count: 516
Tweet ID: 808838249661788160
Tweet Count: 517
Tweet ID: 808733504066486276
Tweet Count: 518
Tweet ID: 808501579447930884
Tweet Count: 519
Tweet ID: 808344865868283904
Tweet Count: 520
Tweet ID: 808134635716833280
Tweet Count: 521
Tweet ID: 808106460588765185
Tweet Count: 522
Tweet ID: 808001312164028416
Tweet Count: 523
Tweet ID: 807621403335917568
Tweet Count: 524
Tweet ID: 807106840509214720
Tweet Count: 525
Tweet ID: 807059379405148160
Tweet Count: 526
Tweet ID: 807010152071229440
Tweet Count: 527
Tweet ID: 806629075125202948
Tweet Count: 528
Tweet ID: 806620845233815552
Tweet Count: 529
Tweet ID: 806576416489959424
Tweet Count: 530
Tweet ID: 806542213899489280
Tweet Count: 531
Tweet ID: 806242860592926720
Tweet Count: 532
Tweet ID: 806219024703037440
Tweet Count: 533
Tweet ID: 8059589

Tweet Count: 688
Tweet ID: 786286427768250368
Tweet Count: 689
Tweet ID: 786233965241827333
Tweet Count: 690
Tweet ID: 786051337297522688
Tweet Count: 691
Tweet ID: 786036967502913536
Tweet Count: 692
Tweet ID: 785927819176054784
Tweet Count: 693
Tweet ID: 785872687017132033
Tweet Count: 694
Tweet ID: 785639753186217984
Tweet Count: 695
Tweet ID: 785533386513321988
Tweet Count: 696
Tweet ID: 785515384317313025
Tweet Count: 697
Tweet ID: 785264754247995392
Tweet Count: 698
Tweet ID: 785170936622350336
Tweet Count: 699
Tweet ID: 784826020293709826
Tweet Count: 700
Tweet ID: 784517518371221505
Tweet Count: 701
Tweet ID: 784431430411685888
Tweet Count: 702
Tweet ID: 784183165795655680
Tweet Count: 703
Tweet ID: 784057939640352768
Tweet Count: 704
Tweet ID: 783839966405230592
Tweet Count: 705
Tweet ID: 783821107061198850
Tweet Count: 706
Tweet ID: 783695101801398276
Tweet Count: 707
Tweet ID: 783466772167098368
Tweet Count: 708
Tweet ID: 783391753726550016
Tweet Count: 709
Tweet ID: 7833475

Tweet Count: 862
Tweet ID: 761227390836215808
Tweet Count: 863
Tweet ID: 761004547850530816
Tweet Count: 864
Tweet ID: 760893934457552897
Tweet Count: 865
Tweet ID: 760656994973933572
Tweet Count: 866
Tweet ID: 760641137271070720
Tweet Count: 867
Tweet ID: 760539183865880579
Tweet Count: 868
Tweet ID: 760521673607086080
Tweet Count: 869
Tweet ID: 760290219849637889
Tweet Count: 870
Tweet ID: 760252756032651264
Tweet Count: 871
Tweet ID: 760190180481531904
Tweet Count: 872
Tweet ID: 760153949710192640
Tweet Count: 873
Tweet ID: 759943073749200896
Tweet Count: 874
Tweet ID: 759923798737051648
Tweet Count: 875
Tweet ID: 759846353224826880
Tweet Count: 876
Tweet ID: 759793422261743616
Tweet Count: 877
Tweet ID: 759566828574212096
Tweet Count: 878
Tweet ID: 759557299618865152
Tweet Count: 879
Tweet ID: 759447681597108224
Tweet Count: 880
Tweet ID: 759446261539934208
Tweet Count: 881
Tweet ID: 759197388317847553
Tweet Count: 882
Tweet ID: 759159934323924993
Tweet Count: 883
Tweet ID: 7590995

Rate limit reached. Sleeping for: 568


An Error occured on ID: 758828659922702336Error Message: Failed to send request: ('Connection aborted.', OSError("(10054, 'WSAECONNRESET')",))
Tweet Count: 886
Tweet ID: 758740312047005698
Tweet Count: 887
Tweet ID: 758474966123810816
Tweet Count: 888
Tweet ID: 758467244762497024
Tweet Count: 889
Tweet ID: 758405701903519748
Tweet Count: 890
Tweet ID: 758355060040593408
Tweet Count: 891
Tweet ID: 758099635764359168
Tweet Count: 892
Tweet ID: 758041019896193024
Tweet Count: 893
Tweet ID: 757741869644341248
Tweet Count: 894
Tweet ID: 757729163776290825
Tweet Count: 895
Tweet ID: 757725642876129280
Tweet Count: 896
Tweet ID: 757611664640446465
Tweet Count: 897
Tweet ID: 757597904299253760
Tweet Count: 898
Tweet ID: 757596066325864448
Tweet Count: 899
Tweet ID: 757400162377592832
Tweet Count: 900
Tweet ID: 757393109802180609
Tweet Count: 901
Tweet ID: 757354760399941633
Tweet Count: 902
Tweet ID: 756998049151549440
Tweet Count: 903
Tweet ID: 756939218950160384
Tweet Count: 904
Tweet ID: 75

Tweet Count: 1058
Tweet ID: 739932936087216128
Tweet Count: 1059
Tweet ID: 739844404073074688
Tweet Count: 1060
Tweet ID: 739623569819336705
Tweet Count: 1061
Tweet ID: 739606147276148736
Tweet Count: 1062
Tweet ID: 739544079319588864
Tweet Count: 1063
Tweet ID: 739485634323156992
Tweet Count: 1064
Tweet ID: 739238157791694849
Tweet Count: 1065
Tweet ID: 738891149612572673
Tweet Count: 1066
Tweet ID: 738885046782832640
Tweet Count: 1067
Tweet ID: 738883359779196928
Tweet Count: 1068
Tweet ID: 738537504001953792
Tweet Count: 1069
Tweet ID: 738402415918125056
Tweet Count: 1070
Tweet ID: 738184450748633089
Tweet Count: 1071
Tweet ID: 738166403467907072
Tweet Count: 1072
Tweet ID: 738156290900254721
Tweet Count: 1073
Tweet ID: 737826014890496000
Tweet Count: 1074
Tweet ID: 737800304142471168
Tweet Count: 1075
Tweet ID: 737678689543020544
Tweet Count: 1076
Tweet ID: 737445876994609152
Tweet Count: 1077
Tweet ID: 737322739594330112
Tweet Count: 1078
Tweet ID: 737310737551491075
Tweet Count: 

Tweet Count: 1233
Tweet ID: 711363825979756544
Tweet Count: 1234
Tweet ID: 711306686208872448
Tweet Count: 1235
Tweet ID: 711008018775851008
Tweet Count: 1236
Tweet ID: 710997087345876993
Tweet Count: 1237
Tweet ID: 710844581445812225
Tweet Count: 1238
Tweet ID: 710833117892898816
Tweet Count: 1239
Tweet ID: 710658690886586372
Tweet Count: 1240
Tweet ID: 710609963652087808
Tweet Count: 1241
Tweet ID: 710588934686908417
Tweet Count: 1242
Tweet ID: 710296729921429505
Tweet Count: 1243
Tweet ID: 710283270106132480
Tweet Count: 1244
Tweet ID: 710272297844797440
Tweet Count: 1245
Tweet ID: 710269109699739648
Tweet Count: 1246
Tweet ID: 710153181850935296
Tweet Count: 1247
Tweet ID: 710140971284037632
Tweet Count: 1248
Tweet ID: 710117014656950272
Tweet Count: 1249
Tweet ID: 709918798883774466
Tweet Count: 1250
Tweet ID: 709901256215666688
Tweet Count: 1251
Tweet ID: 709852847387627521
Tweet Count: 1252
Tweet ID: 709566166965075968
Tweet Count: 1253
Tweet ID: 709556954897764353
Tweet Count: 

Tweet Count: 1408
Tweet ID: 697995514407682048
Tweet Count: 1409
Tweet ID: 697990423684476929
Tweet Count: 1410
Tweet ID: 697943111201378304
Tweet Count: 1411
Tweet ID: 697881462549430272
Tweet Count: 1412
Tweet ID: 697630435728322560
Tweet Count: 1413
Tweet ID: 697616773278015490
Tweet Count: 1414
Tweet ID: 697596423848730625
Tweet Count: 1415
Tweet ID: 697575480820686848
Tweet Count: 1416
Tweet ID: 697516214579523584
Tweet Count: 1417
Tweet ID: 697482927769255936
Tweet Count: 1418
Tweet ID: 697463031882764288
Tweet Count: 1419
Tweet ID: 697270446429966336
Tweet Count: 1420
Tweet ID: 697259378236399616
Tweet Count: 1421
Tweet ID: 697255105972801536
Tweet Count: 1422
Tweet ID: 697242256848379904
Tweet Count: 1423
Tweet ID: 696900204696625153
Tweet Count: 1424
Tweet ID: 696894894812565505
Tweet Count: 1425
Tweet ID: 696886256886657024
Tweet Count: 1426
Tweet ID: 696877980375769088
Tweet Count: 1427
Tweet ID: 696754882863349760
Tweet Count: 1428
Tweet ID: 696744641916489729
Tweet Count: 

Tweet Count: 1583
Tweet ID: 686035780142297088
Tweet Count: 1584
Tweet ID: 686034024800862208
Tweet Count: 1585
Tweet ID: 686007916130873345
Tweet Count: 1586
Tweet ID: 686003207160610816
Tweet Count: 1587
Tweet ID: 685973236358713344
Tweet Count: 1588
Tweet ID: 685943807276412928
Tweet Count: 1589
Tweet ID: 685906723014619143
Tweet Count: 1590
Tweet ID: 685681090388975616
Tweet Count: 1591
Tweet ID: 685667379192414208
Tweet Count: 1592
Tweet ID: 685663452032069632
Tweet Count: 1593
Tweet ID: 685641971164143616
Tweet Count: 1594
Tweet ID: 685547936038666240
Tweet Count: 1595
Tweet ID: 685532292383666176
Tweet Count: 1596
Tweet ID: 685325112850124800
Tweet Count: 1597
Tweet ID: 685321586178670592
Tweet Count: 1598
Tweet ID: 685315239903100929
Tweet Count: 1599
Tweet ID: 685307451701334016
Tweet Count: 1600
Tweet ID: 685268753634967552
Tweet Count: 1601
Tweet ID: 685198997565345792
Tweet Count: 1602
Tweet ID: 685169283572338688
Tweet Count: 1603
Tweet ID: 684969860808454144
Tweet Count: 

Tweet Count: 1755
Tweet ID: 678334497360859136
Tweet Count: 1756
Tweet ID: 678278586130948096
Tweet Count: 1757
Tweet ID: 678255464182861824
Tweet Count: 1758
Tweet ID: 678023323247357953
Tweet Count: 1759
Tweet ID: 678021115718029313
Tweet Count: 1760
Tweet ID: 677961670166224897
Tweet Count: 1761
Tweet ID: 677918531514703872
Tweet Count: 1762
Tweet ID: 677895101218201600
Tweet Count: 1763
Tweet ID: 677716515794329600
Tweet Count: 1764
Tweet ID: 677700003327029250
Tweet Count: 1765
Tweet ID: 677698403548192770
Tweet Count: 1766
Tweet ID: 677687604918272002
Tweet Count: 1767
Tweet ID: 677673981332312066
Tweet Count: 1768
Tweet ID: 677662372920729601
Tweet Count: 1769
Tweet ID: 677644091929329666
Tweet Count: 1770
Tweet ID: 677573743309385728
Tweet Count: 1771
Tweet ID: 677565715327688705
Tweet Count: 1772
Tweet ID: 677557565589463040
Tweet Count: 1773
Tweet ID: 677547928504967168
Tweet Count: 1774
Tweet ID: 677530072887205888
Tweet Count: 1775
Tweet ID: 677335745548390400
Tweet Count: 

Rate limit reached. Sleeping for: 498


An Error occured on ID: 676975532580409345Error Message: Failed to send request: ('Connection aborted.', OSError("(10054, 'WSAECONNRESET')",))
Tweet Count: 1784
Tweet ID: 676957860086095872
Tweet Count: 1785
Tweet ID: 676949632774234114
Tweet Count: 1786
Tweet ID: 676948236477857792
Tweet Count: 1787
Tweet ID: 676946864479084545
Tweet Count: 1788
Tweet ID: 676942428000112642
Tweet Count: 1789
Tweet ID: 676936541936185344
Tweet Count: 1790
Tweet ID: 676916996760600576
Tweet Count: 1791
Tweet ID: 676897532954456065
Tweet Count: 1792
Tweet ID: 676864501615042560
Tweet Count: 1793
Tweet ID: 676821958043033607
Tweet Count: 1794
Tweet ID: 676819651066732545
Tweet Count: 1795
Tweet ID: 676811746707918848
Tweet Count: 1796
Tweet ID: 676776431406465024
Tweet Count: 1797
Tweet ID: 676617503762681856
Tweet Count: 1798
Tweet ID: 676613908052996102
Tweet Count: 1799
Tweet ID: 676606785097199616
Tweet Count: 1800
Tweet ID: 676603393314578432
Tweet Count: 1801
Tweet ID: 676593408224403456
Tweet Count

Tweet Count: 1956
Tweet ID: 673213039743795200
Tweet Count: 1957
Tweet ID: 673148804208660480
Tweet Count: 1958
Tweet ID: 672997845381865473
Tweet Count: 1959
Tweet ID: 672995267319328768
Tweet Count: 1960
Tweet ID: 672988786805112832
Tweet Count: 1961
Tweet ID: 672984142909456390
Tweet Count: 1962
Tweet ID: 672980819271634944
Tweet Count: 1963
Tweet ID: 672975131468300288
Tweet Count: 1964
Tweet ID: 672970152493887488
Tweet Count: 1965
Tweet ID: 672968025906282496
Tweet Count: 1966
Tweet ID: 672964561327235073
Tweet Count: 1967
Tweet ID: 672902681409806336
Tweet Count: 1968
Tweet ID: 672898206762672129
Tweet Count: 1969
Tweet ID: 672884426393653248
Tweet Count: 1970
Tweet ID: 672877615439593473
Tweet Count: 1971
Tweet ID: 672834301050937345
Tweet Count: 1972
Tweet ID: 672828477930868736
Tweet Count: 1973
Tweet ID: 672640509974827008
Tweet Count: 1974
Tweet ID: 672622327801233409
Tweet Count: 1975
Tweet ID: 672614745925664768
Tweet Count: 1976
Tweet ID: 672609152938721280
Tweet Count: 

Tweet Count: 2131
Tweet ID: 669749430875258880
Tweet Count: 2132
Tweet ID: 669684865554620416
Tweet Count: 2133
Tweet ID: 669683899023405056
Tweet Count: 2134
Tweet ID: 669682095984410625
Tweet Count: 2135
Tweet ID: 669680153564442624
Tweet Count: 2136
Tweet ID: 669661792646373376
Tweet Count: 2137
Tweet ID: 669625907762618368
Tweet Count: 2138
Tweet ID: 669603084620980224
Tweet Count: 2139
Tweet ID: 669597912108789760
Tweet Count: 2140
Tweet ID: 669583744538451968
Tweet Count: 2141
Tweet ID: 669573570759163904
Tweet Count: 2142
Tweet ID: 669571471778410496
Tweet Count: 2143
Tweet ID: 669567591774625800
Tweet Count: 2144
Tweet ID: 669564461267722241
Tweet Count: 2145
Tweet ID: 669393256313184256
Tweet Count: 2146
Tweet ID: 669375718304980992
Tweet Count: 2147
Tweet ID: 669371483794317312
Tweet Count: 2148
Tweet ID: 669367896104181761
Tweet Count: 2149
Tweet ID: 669363888236994561
Tweet Count: 2150
Tweet ID: 669359674819481600
Tweet Count: 2151
Tweet ID: 669354382627049472
Tweet Count: 

Tweet Count: 2306
Tweet ID: 666428276349472768
Tweet Count: 2307
Tweet ID: 666421158376562688
Tweet Count: 2308
Tweet ID: 666418789513326592
Tweet Count: 2309
Tweet ID: 666411507551481857
Tweet Count: 2310
Tweet ID: 666407126856765440
Tweet Count: 2311
Tweet ID: 666396247373291520
Tweet Count: 2312
Tweet ID: 666373753744588802
Tweet Count: 2313
Tweet ID: 666362758909284353
Tweet Count: 2314
Tweet ID: 666353288456101888
Tweet Count: 2315
Tweet ID: 666345417576210432
Tweet Count: 2316
Tweet ID: 666337882303524864
Tweet Count: 2317
Tweet ID: 666293911632134144
Tweet Count: 2318
Tweet ID: 666287406224695296
Tweet Count: 2319
Tweet ID: 666273097616637952
Tweet Count: 2320
Tweet ID: 666268910803644416
Tweet Count: 2321
Tweet ID: 666104133288665088
Tweet Count: 2322
Tweet ID: 666102155909144576
Tweet Count: 2323
Tweet ID: 666099513787052032
Tweet Count: 2324
Tweet ID: 666094000022159362
Tweet Count: 2325
Tweet ID: 666082916733198337
Tweet Count: 2326
Tweet ID: 666073100786774016
Tweet Count: 

In [12]:
data

{'tweet': [{'tweet_id': 892420643555336193,
   'retweet_count': 8321,
   'favorite_count': 38042,
   'followers_count': 7595220},
  {'tweet_id': 892177421306343426,
   'retweet_count': 6148,
   'favorite_count': 32668,
   'followers_count': 7595220},
  {'tweet_id': 891815181378084864,
   'retweet_count': 4069,
   'favorite_count': 24577,
   'followers_count': 7595220},
  {'tweet_id': 891689557279858688,
   'retweet_count': 8470,
   'favorite_count': 41407,
   'followers_count': 7595220},
  {'tweet_id': 891327558926688256,
   'retweet_count': 9164,
   'favorite_count': 39575,
   'followers_count': 7595220},
  {'tweet_id': 891087950875897856,
   'retweet_count': 3049,
   'favorite_count': 19871,
   'followers_count': 7595220},
  {'tweet_id': 890971913173991426,
   'retweet_count': 2020,
   'favorite_count': 11619,
   'followers_count': 7595220},
  {'tweet_id': 890729181411237888,
   'retweet_count': 18467,
   'favorite_count': 64205,
   'followers_count': 7595220},
  {'tweet_id': 8906091

In [13]:
# tweet ids which could not be extracted from twitter
tweet_errors.keys()

dict_keys(['888202515573088257', '873697596434513921', '872668790621863937', '869988702071779329', '866816280283807744', '861769973181624320', '845459076796616705', '842892208864923648', '837012587749474308', '827228250799742977', '812747805718642688', '802247111496568832', '775096608509886464', '770743923962707968', '758828659922702336', '754011816964026368', '680055455951884288', '676975532580409345'])

In [14]:
class NumpyEncoder(json.JSONEncoder):
    """ Special json encoder for numpy types """
    def default(self, obj):
        if isinstance(obj, (np.int_, np.intc, np.intp, np.int8,
            np.int16, np.int32, np.int64, np.uint8,
            np.uint16, np.uint32, np.uint64)):
            return int(obj)
        
        return json.JSONEncoder.default(self, obj)

In [15]:
with open('tweet_json.txt', 'w') as json_file:
    for tweet in data['tweet']:
        json.dump(tweet, json_file, cls=NumpyEncoder)
        json_file.write('\n')

https://stackoverflow.com/questions/26646362/numpy-array-is-not-json-serializable/32850511

https://stackoverflow.com/questions/17055117/python-json-dump-append-to-txt-with-each-variable-on-new-line

In [16]:
df3 = pd.read_json('tweet_json.txt', lines=True)

In [17]:
df3.head()

favorite_count  followers_count  retweet_count            tweet_id
0           38042          7595220           8321  892420643555336193
1           32668          7595220           6148  892177421306343426
2           24577          7595220           4069  891815181378084864
3           41407          7595220           8470  891689557279858688
4           39575          7595220           9164  891327558926688256

## Assessing

In [18]:
df1

tweet_id                                            jpg_url  \
0     666020888022790149    https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928    https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449    https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704    https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465    https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
5     666050758794694657    https://pbs.twimg.com/media/CT5Jof1WUAEuVxN.jpg   
6     666051853826850816    https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
7     666055525042405380    https://pbs.twimg.com/media/CT5N9tpXIAAifs1.jpg   
8     666057090499244032    https://pbs.twimg.com/media/CT5PY90WoAAQGLo.jpg   
9     666058600524156928    https://pbs.twimg.com/media/CT5Qw94XAAA_2dP.jpg   
10    666063827256086533    https://pbs.twimg.com/media/CT5Vg_wXIAAXfnj.jpg   
11    666071193221509120    https://pbs.twimg.com/media/CT5cN_3WEAAlOoZ.jpg   
12    666073100786774016    https://pbs.twimg.com/media/CT5d9DZXAAALcwe.jpg   
13    666082916733198337    https://pbs.twimg.com/media/CT5m4VGWEAAtKc8.jpg   
14    666094000022159362    https://pbs.twimg.com/media/CT5w9gUW4AAsBNN.jpg   
15    666099513787052032    https://pbs.twimg.com/media/CT51-JJUEAA6hV8.jpg   
16    666102155909144576    https://pbs.twimg.com/media/CT54YGiWUAEZnoK.jpg   
17    666104133288665088    https://pbs.twimg.com/media/CT56LSZWoAAlJj2.jpg   
18    666268910803644416    https://pbs.twimg.com/media/CT8QCd1WEAADXws.jpg   
19    666273097616637952    https://pbs.twimg.com/media/CT8T1mtUwAA3aqm.jpg   
20    666287406224695296    https://pbs.twimg.com/media/CT8g3BpUEAAuFjg.jpg   
21    666293911632134144    https://pbs.twimg.com/media/CT8mx7KW4AEQu8N.jpg   
22    666337882303524864    https://pbs.twimg.com/media/CT9OwFIWEAMuRje.jpg   
23    666345417576210432    https://pbs.twimg.com/media/CT9Vn7PWoAA_ZCM.jpg   
24    666353288456101888    https://pbs.twimg.com/media/CT9cx0tUEAAhNN_.jpg   
25    666362758909284353    https://pbs.twimg.com/media/CT9lXGsUcAAyUFt.jpg   
26    666373753744588802    https://pbs.twimg.com/media/CT9vZEYWUAAlZ05.jpg   
27    666396247373291520    https://pbs.twimg.com/media/CT-D2ZHWIAA3gK1.jpg   
28    666407126856765440    https://pbs.twimg.com/media/CT-NvwmW4AAugGZ.jpg   
29    666411507551481857    https://pbs.twimg.com/media/CT-RugiWIAELEaq.jpg   
...                  ...                                                ...   
2045  886366144734445568    https://pbs.twimg.com/media/DE0BTnQUwAApKEH.jpg   
2046  886680336477933568    https://pbs.twimg.com/media/DE4fEDzWAAAyHMM.jpg   
2047  886736880519319552    https://pbs.twimg.com/media/DE5Se8FXcAAJFx4.jpg   
2048  886983233522544640    https://pbs.twimg.com/media/DE8yicJW0AAAvBJ.jpg   
2049  887101392804085760    https://pbs.twimg.com/media/DE-eAq6UwAA-jaE.jpg   
2050  887343217045368832  https://pbs.twimg.com/ext_tw_video_thumb/88734...   
2051  887473957103951883    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2052  887517139158093824  https://pbs.twimg.com/ext_tw_video_thumb/88751...   
2053  887705289381826560    https://pbs.twimg.com/media/DFHDQBbXgAEqY7t.jpg   
2054  888078434458587136    https://pbs.twimg.com/media/DFMWn56WsAAkA7B.jpg   
2055  888202515573088257    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2056  888554962724278272    https://pbs.twimg.com/media/DFTH_O-UQAACu20.jpg   
2057  888804989199671297    https://pbs.twimg.com/media/DFWra-3VYAA2piG.jpg   
2058  888917238123831296    https://pbs.twimg.com/media/DFYRgsOUQAARGhO.jpg   
2059  889278841981685760  https://pbs.twimg.com/ext_tw_video_thumb/88927...   
2060  889531135344209921    https://pbs.twimg.com/media/DFg_2PVW0AEHN3p.jpg   
2061  889638837579907072    https://pbs.twimg.com/media/DFihzFfXsAYGDPR.jpg   
2062  889665388333682689    https://pbs.twimg.com/media/DFi579UWsAAatzw.jpg   
2063  889880896479866881    https://pbs.twimg.com/media/DFl99B1WsA

In [19]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [20]:
df2

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
5     891087950875897856                    NaN                  NaN   
6     890971913173991426                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
8     890609185150312448                    NaN                  NaN   
9     890240255349198849                    NaN                  NaN   
10    890006608113172480                    NaN                  NaN   
11    889880896479866881                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
13    889638837579907072                    NaN                  NaN   
14    889531135344209921                    NaN                  NaN   
15    889278841981685760                    NaN                  NaN   
16    888917238123831296                    NaN                  NaN   
17    888804989199671297                    NaN                  NaN   
18    888554962724278272                    NaN                  NaN   
19    888202515573088257                    NaN                  NaN   
20    888078434458587136                    NaN                  NaN   
21    887705289381826560                    NaN                  NaN   
22    887517139158093824                    NaN                  NaN   
23    887473957103951883                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
26    886983233522544640                    NaN                  NaN   
27    886736880519319552                    NaN                  NaN   
28    886680336477933568                    NaN                  NaN   
29    886366144734445568                    NaN                  NaN   
...                  ...                    ...                  ...   
2326  666411507551481857                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
2328  666396247373291520                    NaN                  NaN   
2329  666373753744588802                    NaN                  NaN   
2330  666362758909284353                    NaN                  NaN   
2331  666353288456101888                    NaN                  NaN   
2332  666345417576210432                    NaN                  NaN   
2333  666337882303524864                    NaN                  NaN   
2334  666293911632134144                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   
2336  666273097616637952                    NaN                  NaN   
2337  666268910803644416                    NaN                  NaN   
2338  666104133288665088                    NaN                  NaN   
2339  666102155909144576                    NaN                  NaN   
2340  666099513787052032                    NaN                  NaN   
2341  666094000022159362                    NaN                  NaN   
2342  666082916733198337                    NaN                  NaN   
2343  666073100786774016                    NaN                  NaN   
2344  666071193221509120                    NaN                  NaN   
2345  666063827256086533                    NaN                  NaN   
2346  666058600524156928                    NaN                  NaN   
2347  666057090499244032                    NaN                  NaN   
2348  666055525042405380                    NaN                  NaN   
2349  666051853826850816                    NaN         

In [21]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [22]:
df2.name.value_counts().nlargest(20)

None       745
a           55
Charlie     12
Lucy        11
Cooper      11
Oliver      11
Tucker      10
Penny       10
Lola        10
Bo           9
Winston      9
the          8
Sadie        8
Buddy        7
Toby         7
an           7
Bailey       7
Daisy        7
Dave         6
Oscar        6
Name: name, dtype: int64

In [23]:
df2.rating_denominator.value_counts()

10     2333
11        3
50        3
80        2
20        2
2         1
16        1
40        1
70        1
15        1
90        1
110       1
120       1
130       1
150       1
170       1
7         1
0         1
Name: rating_denominator, dtype: int64

In [24]:
df2.rating_numerator.value_counts().nlargest(15)

12     558
11     464
10     461
13     351
9      158
8      102
7       55
14      54
5       37
6       32
3       19
4       17
1        9
2        9
420      2
Name: rating_numerator, dtype: int64

In [25]:
df3

favorite_count  followers_count  retweet_count            tweet_id
0              38042          7595220           8321  892420643555336193
1              32668          7595220           6148  892177421306343426
2              24577          7595220           4069  891815181378084864
3              41407          7595220           8470  891689557279858688
4              39575          7595220           9164  891327558926688256
5              19871          7595220           3049  891087950875897856
6              11619          7595220           2020  890971913173991426
7              64205          7595220          18467  890729181411237888
8              27313          7595219           4188  890609185150312448
9              31333          7595219           7219  890240255349198849
10             30107          7595219           7178  890006608113172480
11             27302          7595219           4873  889880896479866881
12             47238          7595219           9837  889665388333682689
13             26657          7595219           4442  889638837579907072
14             14841          7595219           2201  889531135344209921
15             24791          7595219           5259  889278841981685760
16             28602          7595219           4404  888917238123831296
17             25106          7595219           4201  888804989199671297
18             19470          7595219           3476  888554962724278272
19             21364          7595219           3419  888078434458587136
20             29640          7595219           5277  887705289381826560
21             45445          7595219          11433  887517139158093824
22             67820          7595219          17792  887473957103951883
23             33097          7595219          10182  887343217045368832
24             30026          7595219           5841  887101392804085760
25             34532          7595219           7612  886983233522544640
26             11841          7595219           3214  886736880519319552
27             22019          7595219           4369  886680336477933568
28             20813          7595219           3135  886366144734445568
29               116          7595219              4  886267009285017600
...              ...              ...            ...                 ...
2308             437          7595332            323  666411507551481857
2309             106          7595332             40  666407126856765440
2310             164          7595332             84  666396247373291520
2311             185          7595332             90  666373753744588802
2312             764          7595332            563  666362758909284353
2313             215          7595332             71  666353288456101888
2314             288          7595332            133  666345417576210432
2315             194          7595332             90  666337882303524864
2316             491          7595332            349  666293911632134144
2317             144          7595332             64  666287406224695296
2318             171          7595332             76  666273097616637952
2319             100          7595332             35  666268910803644416
2320           14095          7595332           6442  666104133288665088
2321              78          7595332             12  666102155909144576
2322             151          7595332             66  666099513787052032
2323             161          7595332             72  666094000022159362
2324             114          7595332             44  666082916733198337
2325             316          7595332            160  666073100786774016
2326             142          7595332             59  666071193221509120
2327             465          7595332            214  666063827256086533
2328             109          7595332             57  666058600524156928
2329             289          7595332            139  666057090499244032
2330             429          7595332            241  6660555250

In [26]:
df3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2338 entries, 0 to 2337
Data columns (total 4 columns):
favorite_count     2338 non-null int64
followers_count    2338 non-null int64
retweet_count      2338 non-null int64
tweet_id           2338 non-null int64
dtypes: int64(4)
memory usage: 73.1 KB


#### Observations

##### Quality Issues:

###### df2 (twitter_archive):
- retweets are included in the df
- replies are included in the df
- value is None instead of 0 for different dog types and name
- timestamp is not a datetime object
- the rating_denominator is not always 10 (should be 10 according to WeRateDogs rating-style-guideline)
- the rating_nominator is not always in range 10 to 14 
- the Dog's name is not always correct

###### df1 (dog breed prediction):
- only 2075 rows (less than in df2) likely because not every tweet included an image
- images where classification is other than dog

###### df3 (query from tweepy api):
- 2338 rows - some tweets did not exist anymore

##### Tidyness Issues:

- doggo, floofer, pupper and puppo are all values of type dog and should therefore be in one column
- the 3 df's should be combined into one

## Cleaning

In [27]:
# create a copy of one df where all the cleaning work will be executed on. (Ensures that there's a backup just in case)
df_clean = df2.copy()

#### Define

Remove all retweets and replies. A tweet can be identified as a reply if the  column ``in_reply_to_status_id`` does contain a value other than NaN. Similarly a tweet is a retweet if the column ``retweeted_status_id`` contains a value other than NaN. The idea therefore is to drop all rows where either of the two columns contains a value other than NaN.

#### Code

In [28]:
df_clean.drop(df_clean[df_clean['in_reply_to_status_id'].notnull()].index, inplace=True)
df_clean.drop(df_clean[df_clean['retweeted_status_id'].notnull()].index, inplace=True)

#### Test

In [29]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2097 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2097 non-null int64
in_reply_to_status_id         0 non-null float64
in_reply_to_user_id           0 non-null float64
timestamp                     2097 non-null object
source                        2097 non-null object
text                          2097 non-null object
retweeted_status_id           0 non-null float64
retweeted_status_user_id      0 non-null float64
retweeted_status_timestamp    0 non-null object
expanded_urls                 2094 non-null object
rating_numerator              2097 non-null int64
rating_denominator            2097 non-null int64
name                          2097 non-null object
doggo                         2097 non-null object
floofer                       2097 non-null object
pupper                        2097 non-null object
puppo                         2097 non-null object
dtypes: float64(4), int64(3), object(10)

The df now contains 2097 rows instead of 2356 and all tweets which were retweets or replies where dropped.

#### Define

The columns ``in_reply_to_status_id``, ``in_reply_to_user_id``, ``retweeted_status_id``, ``retweeted_status_user_id`` and ``retweeted_status_timestamp`` now add no more value to the project, which is why I'm going to drop them now.

#### Code

In [30]:
df_clean.drop(labels=['in_reply_to_status_id',
                       'in_reply_to_user_id', 
                       'retweeted_status_id', 
                       'retweeted_status_user_id', 
                       'retweeted_status_timestamp'], axis=1, inplace=True)

#### Test

In [31]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2097 entries, 0 to 2355
Data columns (total 12 columns):
tweet_id              2097 non-null int64
timestamp             2097 non-null object
source                2097 non-null object
text                  2097 non-null object
expanded_urls         2094 non-null object
rating_numerator      2097 non-null int64
rating_denominator    2097 non-null int64
name                  2097 non-null object
doggo                 2097 non-null object
floofer               2097 non-null object
pupper                2097 non-null object
puppo                 2097 non-null object
dtypes: int64(3), object(9)
memory usage: 213.0+ KB


The columns are now no longer stored in the df.

#### Define

Replace all 'None' values in the columns doggo, floofer, pupper and puppo with actual NaNs. Then combine the 4 indivdual columns to a new column dog_stage whichs contains either the dog stage or NaN. This is achieved with a nested ``.fillna()`` method.

#### Code

In [32]:
dog_stage = ['doggo', 'floofer', 'pupper', 'puppo']

df_clean[dog_stage] = df_clean[dog_stage].replace('None', np.NaN)

In [33]:
df_clean['dog_stage'] = df_clean.doggo.fillna(df_clean.floofer.fillna(df_clean.pupper.fillna(df_clean.puppo)))

#### Test

In [34]:
df_clean.dog_stage.value_counts()

pupper     221
doggo       83
puppo       23
floofer      9
Name: dog_stage, dtype: int64

#### Define

Remove the columns pupper, doggo, puppo and floofer since they are now no longer needed.

#### Code

In [35]:
df_clean.drop(labels=['pupper',
                      'doggo',
                      'puppo',
                      'floofer'], axis=1, inplace=True)

#### Test

In [36]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2097 entries, 0 to 2355
Data columns (total 9 columns):
tweet_id              2097 non-null int64
timestamp             2097 non-null object
source                2097 non-null object
text                  2097 non-null object
expanded_urls         2094 non-null object
rating_numerator      2097 non-null int64
rating_denominator    2097 non-null int64
name                  2097 non-null object
dog_stage             336 non-null object
dtypes: int64(3), object(6)
memory usage: 163.8+ KB


#### Define

Set the dtype of the timestamp column to datetime. This is done using ``pd.to_datetime()``. Add ``dt.normalize()`` to only display the information of year, month and day.

#### Code

In [39]:
df_clean.timestamp = pd.to_datetime(df_clean.timestamp, yearfirst=True).dt.normalize()

#### Test

In [41]:
df_clean.timestamp.sample(3)

292   2017-03-04
370   2017-02-06
862   2016-08-08
Name: timestamp, dtype: datetime64[ns]